### PR TITLE
Topologyobjects coverage

### PR DESCRIPTION
**Pull request by richardjgowers**
#roadto90%

### CODE REVIEW

**Comment by orbeckst** [comment]
(How can I add this PR to the 90% coverage project? https://github.com/MDAnalysis/mdanalysis/projects/2 )

**Comment by dotsdl** [comment]
@orbeckst click the "add cards" button the right, search for the PR, then drag it to the list you want to add it to. I went ahead and did it.

File: testsuite/MDAnalysisTests/core/test_topologyobjects.py [hunk 1]
```diff
@@ -275,6 +279,52 @@ def test_bad_creation_TG(self):
...
+        vals = np.array([[0, 10], [5, 15]])
+
+        tg = TopologyGroup(vals, self.universe, guessed=True)
+
+        assert_array_equal(tg._guessed, np.array([[True], [True]]))
+
+    def test_create_guessed_tg_2(self):
+        vals = np.array([[0, 10], [5, 15]])
+
+        tg = TopologyGroup(vals, self.universe, guessed=False)
```
Comment (dotsdl) [inline testsuite/MDAnalysisTests/core/test_topologyobjects.py]: What does the `guessed` keyword do for `TopologyObject`s? I don't see it documented in the class or `__init__` docs.

File: testsuite/MDAnalysisTests/core/test_topologyobjects.py [hunk 2]
```diff
@@ -430,3 +504,186 @@ def test_TG_getitem_bool_IE(self):
...
+    def test_atom4(self):
+        tg = self.universe.dihedrals[:5]
+
+        a4 = tg.atom4
+        assert_(len(tg) == len(a4))
+        for (atom, bond) in zip(a4, tg):
+            assert_(atom == bond[3])
+
+
+class TestTopologyGroup_Cython(object):
```
Comment (dotsdl) [inline testsuite/MDAnalysisTests/core/test_topologyobjects.py]: I'm not terribly familiar with the bonds/angles/dihedrals machinery here, but wasn't aware at all of any cython bits. What parts are cython?

Comment (richardjgowers) [inline testsuite/MDAnalysisTests/core/test_topologyobjects.py]: Just the raw maths on calculating distances/angles eg: https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/lib/include/calc_distances.h#L334

File: testsuite/MDAnalysisTests/core/test_topologyobjects.py (hunk 1 above)
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/core/test_topologyobjects.py]: It's something from before when I started, it lets people know if a bond was guessed or came from the original topology.

I think the only place it's used elsewhere in the package is writing PDBs, you can choose to only write original CONECT records:

https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/coordinates/PDB.py#L679

File: package/MDAnalysis/core/topologyobjects.py [hunk 3]
```diff
@@ -764,14 +764,50 @@ def __repr__(self):
...
+        try:
+            return self._ags[3]
+        except IndexError:
+            nvert = {'bond': 2,
+                     'angle': 3,
+                     'dihedral': 4,
+                     'improper': 4}[self.btype]
+            raise IndexError("TopologyGroup of {}s only has {} vertical AtomGroups"
+                             "".format(self.btype, nvert))
+
```
Comment (richardjgowers) [inline package/MDAnalysis/core/topologyobjects.py]: It's mostly for the distance calculations.  It just transposes the bond records into vertical slices which is what you want 

```
[(a1, a2, a3), (b1, b2, b3), ...] -> [(a1, b1, ...), (a2, b2, ...), (a3, b3, ...)]
```

So bond length calculation is something like `norm(atom1.positions - atom2.positions)`

**Comment by kain88-de** [comment]
@dotsdl are you happy with the replies of @richardjgowers?

**Comment by richardjgowers** [comment]
@kain88-de can you take a quick look and merge this?  It's mostly just test coverage

File: package/MDAnalysis/core/topologyobjects.py (hunk 3 above)
Comment (kain88-de) [inline package/MDAnalysis/core/topologyobjects.py]: what is so special about the first 4 atoms then? Why do they need a specific attribute and not the other atoms?

Comment (richardjgowers) [inline package/MDAnalysis/core/topologyobjects.py]: We only have Bonds/Angles/Dihedrals, so any connection can only have 4 atoms.

Comment (kain88-de) [inline package/MDAnalysis/core/topologyobjects.py]: I should read the doc strings of classes

**Pull request merged**